### Initial Data - Minimal Edits ###
This is the initial dataset formatted for pandas dataframe <br/>
This notebook is strictly for data preprocessing

##### What This Notebook Will Do #####
- Remove NFR requirements classes where there is too little data or that class is not need (Functional)
- Remove all non-unique sentences
- Remove all non alphabetical characters
- Remove all 'common' stop-words
- Export final class to new .txt file

This new .txt will be manually examined for any remaining errors: grammatical, spelling or otherwise

A later decision will be made to remove any custom stop-words in an effort to improve the model's accuracy


In [87]:
import pandas as pd

#putting correctly formatted data in pandas DataFrame
filename = 'DataFiles/InitialData.txt'
df = pd.read_csv(filename)
grouped_classes = df.groupby(['class'])
grouped_classes.describe()

sentence         \
                            count unique   
class                                      
access control                473    464   
audit                          99     99   
availability                   38     38   
capacity and performance       74     73   
database design               965    959   
functional                   1793   1787   
legal                          66     66   
look and feel                  67     67   
maintainability               133    133   
operational                   101    101   
other nonfunctional            38     38   
privacy                       247    246   
recoverability                 30     30   
reliability                    22     22   
security                      304    302   
usability                     107    106   

                                                                             \
                                                                        top   
class                                                                         
access control            The LHCP can select a patient to obtain additi...   
audit                       The irregular activity reports are customizable   
availability              The product shall be available during normal b...   
capacity and performance  The response time shall be fast enough to main...   
database design           The system shall provide the ability to save a...   
functional                Upon reading the report the public health agen...   
legal                     Recipient shall not attempt to identify the in...   
look and feel             All screens created as part of the Disputes ap...   
maintainability           Other law (including regulations adopted by th...   
operational               The product shall display HTML properly in 80%...   
other nonfunctional       The Proposer shall adhere to service levels de...   
privacy                   The product shall provide players no access to...   
recoverability            The Proposer shall provide a recovery environm...   
reliability               The website shall continue to operate if the p...   
security                  Protect electronic health information created ...   
usability                 The product shall be intuitive and self-explan...   

                               
                         freq  
class                          
access control              2  
audit                       1  
availability                1  
capacity and performance    2  
database design             2  
functional                  2  
legal                       1  
look and feel               1  
maintainability             1  
operational                 1  
other nonfunctional         1  
privacy                     2  
recoverability              1  
reliability                 1  
security                    2  
usability                   2

### Trimming the data ###
The data set is clearly imbalanced so we are going to trim it down <br />
###### Classes that we will include: ######
- Access Control
- Database Design
- Privacy
- Security

In [88]:
import re 
import io

#import nltk 
#nltk.download('stopwords')
#from nltk.corpus import stopwords
#not sure if we should use the NLTK corpus....has some weird entries in it
#print(stopwords.words('english'))


removal_array = ['audit', 'availability', 'capacity and performance', 'functional', 'legal', 'look and feel', 
                'maintainability','reliability','operational', 'other nonfunctional', 'recoverability', 'reliability', 
                 'usability']

#loop through the removal_array
#remove the the row for that specific class value from the data frame
for i in removal_array:
    df = df[df['class'] != i]
    
#drop all the duplicate rows
df = df.drop_duplicates()

#remove all non-alphabetical characters 
#make all words lower case
for ind in df.index:
    current_sentence = df['sentence'][ind]
    current_sentence = re.sub('[^a-zA-Z_ ]+', '', current_sentence)
    lower = current_sentence.lower()
    df['sentence'][ind] = lower

    #Stop words and custom stop phrases
stop_words = ['the ', ' the ' ' and ', ' or ', ' a ', ' an ',  ' of ', ' to ', ' that ', ' is ', 
              ' by ', ' you ', ' not ', ' can ', ' user ', ' when ', ' are ', ' this ', ' who ', 'whom', 'if']
custom_stop_phrases = ['the system shall ', 'the product shall ', 
                    ' provide the ability ', ' have the ability ', ' has the capability ', ' has the ability '] 

# * works like the ... (spread) operator from javascript
full_stop = [*custom_stop_phrases, *stop_words]

#remove stop items from each sentence in Dataframe
for ind in df.index:
    for stop in full_stop:
        df['sentence'][ind] = df['sentence'][ind].replace(stop, ' ')

df.to_csv('DataFiles/PostProcessing.txt', encoding='utf-8', index=False)



#### Data Trimming ####
The above data will be found in /DataFiles/PostProcessing.txt <br />
The data will be trimmed to have the following sizes <br />

- Access Control - 350
- Database Design - 350
- Privacy - 246
- Security - 302

In [89]:
new_classes = df.groupby(['class'])
new_classes.describe()

sentence         \
                   count unique   
class                             
access control       464    464   
database design      959    958   
privacy              246    246   
security             302    302   

                                                                         
                                                               top freq  
class                                                                    
access control      chooses open his her upcoming appointment list    1  
database design  generate lists patients specific conditions us...    2  
privacy          a covered health care provider may condition  ...    1  
security         system provides integrated security managed in...    1